File: packages/core/src/algorithms/neighbor-aggregate.ts
import type { Edge } from '../core/Edge.js';
import type { Graph } from '../core/Graph.js';
import type { Vertex } from '../core/Vertex.js';
import { type AlgorithmGen, defineAlgorithm, YIELD_EVERY } from './async.js';
import type { AlgorithmConfig, AlgorithmRow } from './types.js';

/** A neighbour-aggregation result row: `{ node, vector }` (the aggregated vector).
 * The column is `vector` (not `aggregate`, which is a reserved GQL word). */
export type NeighborAggregateRow = AlgorithmRow<'vector', number[]>;

type Op = 'mean' | 'sum' | 'max' | 'min';

/** Read a vertex's feature as a numeric vector, or `null` if absent / not a list of
 * numbers (such a vertex contributes nothing). Mirrors native `read_vec`. */
const readVec = (vertex: Vertex, key: string): number[] | null => {
  const raw = vertex.getProperty(key);

  if (!Array.isArray(raw)) {
    return null;
  }

  const out: number[] = [];

  for (const x of raw) {
    if (typeof x !== 'number') {
      return null; // a non-numeric element → not a feature vector
    }

    out.push(x);
  }

  return out;
};

/** Fold one contributor vector into `acc` under `op`, scaled by `coef` (edge weight ×
 * normalization; 1 unweighted). `started` marks whether `acc` holds a real value yet (the
 * identity for max/min is the first contributor). `max`/`min` ignore `coef` (scale-
 * independent; a weight/norm is rejected at the call site). Mirrors native `fold`. */
const fold = (op: Op, acc: number[], vec: number[], started: boolean, coef: number): void => {
  for (let i = 0; i < acc.length; i++) {
    if (op === 'sum' || op === 'mean') {
      acc[i] += coef * vec[i];
    } else if (!started) {
      acc[i] = vec[i];
    } else if (op === 'max') {
      acc[i] = Math.max(acc[i], vec[i]);
    } else {
      acc[i] = Math.min(acc[i], vec[i]);
    }
  }
};

type Contributor = { eidx: number; nbr: string };

/** Validate `op` (raw so the message shows the offending value, not `never`). */
const resolveOp = (raw: string): Op => {
  if (raw !== 'mean' && raw !== 'sum' && raw !== 'max' && raw !== 'min') {
    throw new Error(`neighborAggregate \`op\` must be one of mean|sum|max|min, got '${raw}'`);
  }

  return raw;
};

/** Validate `direction` into `[wantOut, wantIn]`. */
const resolveDirs = (raw: string): [boolean, boolean] => {
  if (raw !== 'out' && raw !== 'in' && raw !== 'both') {
    throw new Error(`neighborAggregate \`direction\` must be one of out|in|both, got '${raw}'`);
  }

  return [raw === 'out' || raw === 'both', raw === 'in' || raw === 'both'];
};

/** Validate `norm` into a GCN flag. */
const resolveGcn = (raw: string): boolean => {
  if (raw !== 'none' && raw !== 'gcn') {
    throw new Error(`neighborAggregate \`norm\` must be one of none|gcn, got '${raw}'`);
  }

  return raw === 'gcn';
};

/** Per-edge weights by edge-insertion index (= native eidx); a missing / non-numeric
 * value is 0, mirroring native `edge_weights`. */
const buildEdgeWeights = (graph: Graph, key: string): number[] => {
  const out: number[] = [];

  for (const edge of graph.edges) {
    const w = edge.getProperty(key);
    out.push(typeof w === 'number' ? w : 0);
  }

  return out;
};

/** Every vertex's feature vector (by id), plus the shared dimension `d`; a length
 * mismatch faults. Iterated in insertion (= dense-id) order. */
const buildFeatures = (
  graph: Graph,
  feature: string,
): { feats: Map<string, number[] | null>; d: number } => {
  const feats = new Map<string, number[] | null>();
  let dim: number | undefined;

  for (const vertex of graph.vertices) {
    const vec = readVec(vertex, feature);

    feats.set(vertex.id, vec);

    if (vec !== null && dim === undefined) {
      dim = vec.length;
    } else if (vec !== null && dim !== vec.length) {
      throw new Error(
        `neighborAggregate feature vectors must all have the same length; found ${dim} and ${vec.length}`,
      );
    }
  }

  return { feats, d: dim ?? 0 };
};

/** Per-vertex out/in adjacency in edge-insertion (= native eidx) order. */
const buildAdjacency = (
  graph: Graph,
  edgeLabel: string | undefined,
): { outAdj: Map<string, Contributor[]>; inAdj: Map<string, Contributor[]> } => {
  const outAdj = new Map<string, Contributor[]>();
  const inAdj = new Map<string, Contributor[]>();
  const typeOk = (edge: Edge): boolean => edgeLabel === undefined || edge.labels.has(edgeLabel);
  let eidx = 0;

  for (const edge of graph.edges) {
    if (typeOk(edge)) {
      const from = edge.from.id;
      const to = edge.to.id;
      (outAdj.get(from) ?? outAdj.set(from, []).get(from)!).push({ eidx, nbr: to });
      (inAdj.get(to) ?? inAdj.set(to, []).get(to)!).push({ eidx, nbr: from });
    }

    eidx++;
  }

  return { outAdj, inAdj };
};

/** Contributors of `v` by direction, sorted by edge index (the canonical,
 * engine-independent accumulation order). A `both`-direction self-loop counts once. */
const gatherContributors = (
  v: string,
  outAdj: Map<string, Contributor[]>,
  inAdj: Map<string, Contributor[]>,
  wantOut: boolean,
  wantIn: boolean,
): Contributor[] => {
  const contrib: Contributor[] = [];

  if (wantOut) {
    contrib.push(...(outAdj.get(v) ?? []));
  }

  if (wantIn) {
    for (const a of inAdj.get(v) ?? []) {
      if (!(wantOut && a.nbr === v)) {
        contrib.push(a);
      }
    }
  }

  return contrib.sort((x, y) => x.eidx - y.eidx);
};

export const computeGen = function* (
  config: AlgorithmConfig,
  graph: Graph,
): AlgorithmGen<NeighborAggregateRow> {
  const { feature, edgeLabel, direction = 'both', writeProperty, weightProperty } = config;

  if (feature === undefined) {
    throw new Error('neighborAggregate requires a `feature` property');
  }

  const op = resolveOp(config.op ?? 'mean');
  const [wantOut, wantIn] = resolveDirs(direction);
  const includeSelf = config.includeSelf ?? false;
  const gcn = resolveGcn(config.norm ?? 'none');
  // A weight or `gcn` norm SCALES contributors — meaningless for order/scale-independent
  // max/min. Reject loudly rather than silently ignore. Mirrors native.
  const weighted = weightProperty !== undefined;

  if ((weighted || gcn) && (op === 'max' || op === 'min')) {
    throw new Error(
      'neighborAggregate `weightProperty`/`norm` apply only to op=sum|mean (max/min are scale-independent)',
    );
  }

  const { feats, d } = buildFeatures(graph, feature);
  const { outAdj, inAdj } = buildAdjacency(graph, edgeLabel);
  const edgeWeights = weighted ? buildEdgeWeights(graph, weightProperty) : [];

  // GCN degree per vertex: contributor count under the configured direction/etype (+ the
  // self-loop when includeSelf), floored at 1. Mirrors native `deg`.
  const deg = new Map<string, number>();

  if (gcn) {
    for (const vertex of graph.vertices) {
      const c = gatherContributors(vertex.id, outAdj, inAdj, wantOut, wantIn).length;
      deg.set(vertex.id, Math.max(c + (includeSelf ? 1 : 0), 1));
    }
  }

  const rows: NeighborAggregateRow[] = [];
  let sinceYield = 0;

  for (const vertex of graph.vertices) {
    const v = vertex.id;
    const acc = new Array<number>(d).fill(0);
    let coefSum = 0;
    let started = false;

    // Self first (when included), with GCN factor 1/deg_i (= 1/sqrt(deg_i·deg_i)).
    if (includeSelf) {
      const sv = feats.get(v) ?? null;

      if (sv !== null) {
        const coef = gcn ? 1 / deg.get(v)! : 1;
        fold(op, acc, sv, started, coef);
        started = true;
        coefSum += coef;
      }
    }

    // Then neighbours in edge-index order, each scaled by edge weight × GCN factor.
    for (const { eidx, nbr } of gatherContributors(v, outAdj, inAdj, wantOut, wantIn)) {
      const nv = feats.get(nbr) ?? null;

      if (nv !== null) {
        const w = weighted ? edgeWeights[eidx] : 1;
        const nf = gcn ? 1 / Math.sqrt(deg.get(v)! * deg.get(nbr)!) : 1;
        const coef = w * nf;
        fold(op, acc, nv, started, coef);
        started = true;
        coefSum += coef;
      }
    }

    if (op === 'mean' && coefSum !== 0) {
      for (let i = 0; i < d; i++) {
        acc[i] /= coefSum;
      }
    }

    if (writeProperty !== undefined) {
      vertex.setProperty(writeProperty, acc);
    }

    rows.push({ node: vertex.id, vector: acc });

    if (++sinceYield >= YIELD_EVERY) {
      sinceYield = 0;

      yield;
    }
  }

  return rows;
};

/**
 * Vectorized neighbour aggregation — for each vertex, aggregate its neighbours'
 * list-valued `feature` vectors element-wise over the whole block (`op` = mean /
 * sum / max / min), by `direction`, optionally including the vertex's own vector.
 * Byte-identical to native `neighborAggregate`: contributors fold in ascending
 * edge-index order (self first when included). Data-last dual-form.
 */
export const neighborAggregate = defineAlgorithm(computeGen);

File: crates/lenke-core/src/algo/neighbor_aggregate.rs
//! Vectorized neighbour aggregation (`neighborAggregate` / message passing).
//!
//! For each vertex, aggregate the list-valued `feature` vectors of its neighbours
//! **element-wise over the whole D-dim block in one native pass** — the primitive
//! host-driven GCN / GraphSAGE message passing wants (instead of D separate GQL
//! `SET`s). `op` chooses `mean` (default) / `sum` / `max` / `min`; `direction`
//! picks out- / in- / both-neighbours; `includeSelf` adds the vertex's own vector.
//! Each contributor is scaled by a COEFFICIENT = edge weight (`weightProperty`, 1.0
//! unweighted) × normalization (`norm:"gcn"` → `1/sqrt(deg_i·deg_j)`, else 1.0), so
//! `sum` = Σ coefⱼ·hⱼ and `mean` = that ÷ Σ coefⱼ (a WEIGHTED mean). A weight/norm is
//! rejected for `max`/`min` (scale-independent). The result is a list value per vertex,
//! optionally written to `writeProperty`.
//!
//! **Byte-identity with the TS mirror** rests on a fixed accumulation order: the
//! vertex's own vector first (when included), then each neighbour in ascending
//! **edge-index** order — so the f64 `sum`/`mean` adds land in the same order in both
//! engines. `max`/`min` are order-independent; degrees are integers (exact) so the GCN
//! `1/sqrt(deg_i·deg_j)` matches. A `both`-direction self-loop is counted once (the
//! in-side copy is dropped, mirroring `expand`).

use crate::algo::AlgoConfig;
use crate::graph::{Graph, Value};

#[derive(Clone, Copy, PartialEq)]
enum Op {
    Mean,
    Sum,
    Max,
    Min,
}

/// Read vertex `v`'s `key` property as a numeric vector, or `None` if it is absent
/// or not a list of numbers (such a vertex contributes nothing to an aggregate).
/// A typed `Column::Vec` feature is a zero-copy slice (no boxing) → the fast path;
/// a list still boxed in a `Mixed` column falls back to unboxing via `value`.
fn read_vec(graph: &Graph, v: u32, key: &str) -> Option<Vec<f64>> {
    if let Some(slice) = graph.props.vector(v as usize, key) {
        return Some(slice.to_vec()); // contiguous copy — no per-element unboxing
    }
    match graph.props.value(v as usize, key, &graph.strs) {
        Value::List(items) => {
            let mut out = Vec::with_capacity(items.len());
            for it in &items {
                match it {
                    Value::Num(n) => out.push(*n),
                    _ => return None, // a non-numeric element → not a feature vector
                }
            }
            Some(out)
        }
        _ => None,
    }
}

/// Fold one contributor vector into `acc` under `op`, scaled by `coef` (the edge
/// weight × normalization factor; `1.0` for a plain unweighted aggregate). `coef_sum`
/// accumulates the coefficients (the denominator for a weighted `mean`); `started`
/// tracks whether `acc` holds a real value yet (for `max`/`min`, whose identity is the
/// first contributor, not zero). `max`/`min` ignore `coef` — they are scale-independent
/// and reject a weight/norm at the call site.
fn fold(op: Op, acc: &mut [f64], coef_sum: &mut f64, started: &mut bool, vec: &[f64], coef: f64) {
    match op {
        Op::Sum | Op::Mean => {
            for (a, x) in acc.iter_mut().zip(vec) {
                *a += coef * x;
            }
        }
        Op::Max => {
            if *started {
                for (a, x) in acc.iter_mut().zip(vec) {
                    *a = a.max(*x);
                }
            } else {
                acc.copy_from_slice(vec);
            }
        }
        Op::Min => {
            if *started {
                for (a, x) in acc.iter_mut().zip(vec) {
                    *a = a.min(*x);
                }
            } else {
                acc.copy_from_slice(vec);
            }
        }
    }
    *started = true;
    *coef_sum += coef;
}

pub fn neighbor_aggregate(graph: &Graph, cfg: &AlgoConfig) -> Result<Vec<(u32, Value)>, String> {
    let feature = cfg
        .feature
        .as_deref()
        .ok_or_else(|| "neighborAggregate requires a `feature` property".to_string())?;
    let op = match cfg.op.as_deref().unwrap_or("mean") {
        "mean" => Op::Mean,
        "sum" => Op::Sum,
        "max" => Op::Max,
        "min" => Op::Min,
        other => {
            return Err(format!(
                "neighborAggregate `op` must be one of mean|sum|max|min, got '{other}'"
            ));
        }
    };
    let (want_out, want_in) = match cfg.direction.as_deref().unwrap_or("both") {
        "out" => (true, false),
        "in" => (false, true),
        "both" => (true, true),
        other => {
            return Err(format!(
                "neighborAggregate `direction` must be one of out|in|both, got '{other}'"
            ));
        }
    };
    let include_self = cfg.include_self.unwrap_or(false);
    let gcn = match cfg.norm.as_deref().unwrap_or("none") {
        "none" => false,
        "gcn" => true,
        other => {
            return Err(format!(
                "neighborAggregate `norm` must be one of none|gcn, got '{other}'"
            ));
        }
    };
    // Per-edge weights (`None` = unweighted, coefficient 1.0). A weight or a `gcn` norm
    // SCALES each contributor, which is meaningless for the order/scale-independent
    // `max`/`min` — reject loudly rather than silently ignore.
    let weighted = cfg.weight_property.is_some();
    if (weighted || gcn) && matches!(op, Op::Max | Op::Min) {
        return Err(
            "neighborAggregate `weightProperty`/`norm` apply only to op=sum|mean \
             (max/min are scale-independent)"
                .to_string(),
        );
    }
    let weights: Option<Vec<f64>> = cfg
        .weight_property
        .as_deref()
        .map(|k| crate::algo::edge_weights(graph, k));
    // `Some(None)` = every type, `Some(Some(id))` = one, `None` = named-but-unknown
    // (no edges match, so every aggregate is over an empty neighbourhood).
    let etype = cfg.etype(graph);

    // Precompute every vertex's feature vector once (avoids re-reading/re-allocating
    // per neighbour) and infer the shared dimension `d`; a length mismatch faults.
    let feats: Vec<Option<Vec<f64>>> = graph
        .vertex_indices()
        .map(|v| read_vec(graph, v, feature))
        .collect();
    let mut dim: Option<usize> = None;
    for f in feats.iter().flatten() {
        match dim {
            None => dim = Some(f.len()),
            Some(d) if d != f.len() => {
                return Err(format!(
                    "neighborAggregate feature vectors must all have the same length; found {} and {}",
                    d,
                    f.len()
                ));
            }
            _ => {}
        }
    }
    let d = dim.unwrap_or(0);

    let etype_ok = |t: u32| match etype {
        Some(None) => true,
        Some(Some(id)) => t == id,
        None => false,
    };

    // Gather a vertex's contributor `(eidx, nbr)` pairs by direction, sorted by edge
    // index for a canonical, engine-independent accumulation order. A `both`-direction
    // self-loop is counted once (the in-side copy is dropped, mirroring `expand`).
    let contributors = |v: u32| -> Vec<(u32, u32)> {
        let mut contrib: Vec<(u32, u32)> = Vec::new();
        if want_out {
            for a in graph.out_adj(v) {
                if etype_ok(a.etype) {
                    contrib.push((a.eidx, a.nbr));
                }
            }
        }
        if want_in {
            for a in graph.in_adj(v) {
                if want_out && a.nbr == v {
                    continue;
                }
                if etype_ok(a.etype) {
                    contrib.push((a.eidx, a.nbr));
                }
            }
        }
        contrib.sort_unstable_by_key(|&(eidx, _)| eidx);
        contrib
    };

    // GCN degree per vertex: its contributor count under the SAME direction/etype filter,
    // plus the self-loop when `includeSelf` (the `Ã = A + I` self-loop). Floored at 1 so a
    // sink/source contributes a finite `1/sqrt(deg_i·deg_j)` instead of dividing by zero.
    let deg: Vec<f64> = if gcn {
        graph
            .vertex_indices()
            .map(|v| (contributors(v).len() + usize::from(include_self)).max(1) as f64)
            .collect()
    } else {
        Vec::new()
    };

    let mut out: Vec<(u32, Value)> = Vec::with_capacity(graph.vertex_count());
    for v in graph.vertex_indices() {
        let contrib = contributors(v);

        let mut acc = vec![0.0f64; d];
        let mut coef_sum = 0.0f64;
        let mut started = false;
        // The coefficient scaling a contributor at edge `eidx` from neighbour `nbr`:
        // edge weight (1.0 unweighted) × GCN factor (`1/sqrt(deg_i·deg_j)`, else 1.0).
        let coef_of = |eidx: u32, nbr: u32| -> f64 {
            let w = weights.as_ref().map_or(1.0, |ws| ws[eidx as usize]);
            let nf = if gcn {
                1.0 / (deg[v as usize] * deg[nbr as usize]).sqrt()
            } else {
                1.0
            };
            w * nf
        };
        if include_self {
            if let Some(sv) = &feats[v as usize] {
                // The self-loop has weight 1.0 and GCN factor `1/deg_i` (`sqrt(deg_i·deg_i)`).
                let coef = if gcn { 1.0 / deg[v as usize] } else { 1.0 };
                fold(op, &mut acc, &mut coef_sum, &mut started, sv, coef);
            }
        }
        for &(eidx, nbr) in &contrib {
            if let Some(nv) = &feats[nbr as usize] {
                fold(
                    op,
                    &mut acc,
                    &mut coef_sum,
                    &mut started,
                    nv,
                    coef_of(eidx, nbr),
                );
            }
        }
        if op == Op::Mean && coef_sum != 0.0 {
            for a in &mut acc {
                *a /= coef_sum;
            }
        }
        // No contributors → the zero vector (acc is already zeros; `max`/`min` also
        // leave zeros since `started` stayed false).
        out.push((v, Value::List(acc.into_iter().map(Value::Num).collect())));
    }
    Ok(out)
}
